# Clustering analysis of transposable element expression

This repository contains a pipeline that performs clustering on mixed human gene/transposable element RNA-sequencing data. 

As an example, we include an RNA-sequencing data set published by 
[Blanco-Melo et al (2020)](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE147507). Transposable element read count data were generated by applying 
[Telescope](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1006453) to the raw read data generated 
by Blanco-Melo and colleagues. A human/HERV gene correlation network is constructed and clusters of HERVs/human genes are produced by applying hierarchical clustering with a 
[dynamic tree cut](https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/BranchCutting/).

## Dependencies

This package uses R version 4.0. All required packages are installed automatically by running the scripts (if not already present).

## Running the pipeline

The pipeline consists of four scripts, each with a unique set of arguments. If starting an analysis, these should be run in order. The four scripts (and their command line arguments) are:

1. [`differentialexpression.R`](differentialexpression.R) - identifies differentially-expressed genes with DESeq2
    * `--human_counts` : filepath to human count matrix (tab-separated)
    * `--TE_counts` : filepath to TE count matrix (tab-separated)
    * `--metadata` : filepath to metadata file (comma-separated), with three columns: "sample", "condition", and "batch" (batch can be omitted)
    * `--compare` : filepath to comparison file (comma-separated), with two columns "A" and "B" for comparisons to be performed by DESeq2 (A = control, B = treatment). Values within each column A/B should be levels of "condition" in the metadata file.
    * `--outdir` : path to directory in which to save data files (this argument can be omitted, in which case data files will be saved to the current directory)
2. [`clustering.R`](clustering.R) - runs average-linkage hierachical clustering on the combined human/TE read count data 
    * `--counts` : filepath to the counts object `counts.Rdata` outputted by [`differentialexpression.R`](differentialexpression.R)
    * `--diffexp` : filepath to the differential expression object `DiffExp.Rdata` outputted by [`differentialexpression.R`](differentialexpression.R)
    * `--geneset` : one of (LRT, All) defining the gene set to use in clustering. LRT only clusters genes/TEs with adj p < 0.05 from the likelihood ratio test. Defaults to LRT.
    * `--minclustersize` : integer specifying the minimum cluster size (default is 200, can be omitted)
    * `--outdir` : directory to save data files (this argument can be omitted, in which case data files will be saved to the current directory)
3. [`plotclusters.R`](plotclusters.R) - plots expression levels within each cluster
    * `--clusterdata` : filepath to the cluster data object `ClusterData.Rdata` outputted by [`clustering.R`](clustering.R)
    * `--metadata` : filepath to metadata file (comma-separated), with three columns: "sample", "condition", and "batch" (batch can be omitted)
    * `--outdir` : directory to save data files (this argument can be omitted, in which case data files will be saved to the current directory)
4. [`clusterenrichment.R`](clusterenrichment.R) - conducts enrichment analysis for a specified set of clusters, plots in heatmap
    * `--clusterdata` : filepath to the cluster data object `ClusterData.Rdata` outputted by [`clustering.R`](clustering.R)
    * `--whichclusters` : space-separated values indicating which clusters enriched terms should be plotted for (e.g. --whichclusters 1 2 3 4 plots clusters 1, 2, 3, and 4), note that the user can include an arbitrary number of clusters here (not just 4)
    * `--outdir` : directory to save data files (this argument can be omitted, in which case data files will be saved to the current directory)
    
As an example, to execute the differential expression script with the Calu3 cell data set included in this directory, the command is:

```
Rscript differentialexpression.R --human_counts data/Gene_Counts_Calu3.txt --TE_counts data/Retro_Counts_Calu3.txt --metadata data/Calu3_metadata.csv --compare data/Calu3_compare.csv --outdir data
```

Which would output the relevant data files to the `data/` directory. After running the differential expression script, you would then run the clustering:

```
Rscript clustering.R --counts data/counts.Rdata --diffexp data/DiffExp.Rdata --geneset LRT --minclustersize 200 --outdir data
```

Then, to plot the clusters:

```
Rscript plotclusters.R --clusterdata data/ClusterData.Rdata --metadata data/Calu3_metadata.csv --outdir data
```

And finally to run the enrichment analysis:

```
Rscript clusterenrichment.R --clusterdata data/ClusterData.Rdata --whichclusters 3 4 10 11 --outdir data
```

We also include another data set of A549 cells in the `data/` directory.